wp-content/civi-extensions/civiglific/CRM/Civiglific/Page/GroupMapping.php

```diff
@@ -44,15 +43,74 @@ public function run() {
 
       $body = $response->getBody()->getContents();
       $data = json_decode($body, TRUE);
-      $groups = $data['data']['groups'] ?? [];
+      $glificGroups = $data['data']['groups'] ?? [];
 
     }
     catch (RequestException $e) {
       CRM_Core_Error::debug_log_message('Glific API HTTP error: ' . $e->getMessage());
-      $groups = [];
+      $glificGroups = [];
+    }
+
+    // Fetch CiviCRM groups.
+    $rawGroups = Group::get(TRUE)
+      ->addSelect('id', 'title')
+      ->addWhere('is_active', '=', 1)
+      ->execute();
+
+    $civicrmGroups = [];
+    foreach ($rawGroups as $group) {
+      $civicrmGroups[$group['id']] = $group['title'];
+    }
+
+    // Handle form submission.
+    if (!empty($_POST['add_rule'])) {
+      $civiGroupId = (int) $_POST['civicrm_group_id'];
+      $glificGroupId = (int) $_POST['glific_group_id'];
+
+      if ($civiGroupId && $glificGroupId) {
+        try {
+          GlificGroupMap::create()
+            ->addValue('group_id', $civiGroupId)
+            ->addValue('collection_id', $glificGroupId)
+            ->addValue('last_sync_date', date('Y-m-d H:i:s'))
+            ->execute();
+
+          $this->assign('success_message', 'New rule added successfully.');
+        }
+        catch (Exception $e) {
+          CRM_Core_Error::debug_log_message('Error creating Glific Group Map: ' . $e->getMessage());
+          $this->assign('error_message', 'Failed to add rule: ' . $e->getMessage());
+        }
+      }
+      else {
+        $this->assign('error_message', 'Please select both groups.');
+      }
```

Comment:
This controller method is doing too much. Consider refactoring into separate helper methods

Currently, `run()` is:
- Handling API token fetch
- Processing form submission
- Reading from DB
- Writing to DB
- Assigning view variables

This breaks the SRP (Single Responsibility Principle), and makes the code hard to test or extend. I’d recommend extracting:
- `fetchGlificGroups()`
- `fetchCiviGroups()`
- `handleRuleSubmission()`
- `getExistingMappings()`

This would help isolate business logic from presentation glue and make the page easier to evolve (e.g. adding AJAX, permissions, or validations).